lib/LaTeXML/MathParser.pm

```diff
@@ -684,11 +685,62 @@ sub parse_single {
       $result = ['ltx:XMDual', {},
         LaTeXML::Package::createXMRefs($document, $result),
         ['ltx:XMWrap', {}, $result, @punct]]; }    # or perhaps: Apply, punctuated???
-    if ($LaTeXML::MathParser::DEBUG) {
+    if ($LaTeXML::MathParser::DEBUG && @$LaTeXML::MathParser::RuleTree) {
+      my $rule_tree = $LaTeXML::MathParser::RuleTree;
+      # First rule is a single item, use for setup:
+      my $top    = shift @$rule_tree;
+      my @levels = ([named_rules_filter($$top[2])]);
+      my @rules  = @{ $$top[3] };
+      while (@rules) {
+        my @subrules = ();
+        # expand the last level
+        my @level = @{ $levels[-1] };
+        # with this level's productions
+        for my $rule (@rules) {
+          my $prev = $$rule[1];
+          my $next = $$rule[2];
+          push @subrules, @{ $$rule[3] };
+          my @level_swap     = ();
+          my $lvl_rule_count = scalar(@level);
+          while (my $lvl_rule = shift @level) {
+            if ($lvl_rule eq $prev) {
+              my @next = named_rules_filter($next);
+              push @level_swap, '(', @next, ')' if @next;
+              last; }
+            else {
+              push @level_swap, $lvl_rule; } }
+          if (scalar(@level_swap) + scalar(@level) == $lvl_rule_count) {
+            ### IMPORTANT: The auto-actions won't fire for "(s)" modified rules, such as "punctExpr(s)".
+            print STDERR "Failed to attach: $prev:$next\n\tat ", join(" ", @level_swap), "\n"; }
```

Comment:
There are actually quite many of these "failed to attach" cases, because the tracing code is really trying to ride on the back of RecDescent uninvited, and every time a partial parse failure occurs, RecDescent can give the tracing logic a slip -- as I can't foresee all possible kinds of derivation jumps the parser makes. I'm not saying it's impossible to do in general, but it already feels prohibitively fragile/dirty from what I've tried.